tardis/io/model/readers/artis.py

```diff
@@ -49,16 +87,104 @@ def read_artis_density(fname):
+        [(28, 56), (27, 56), (26, 52), (24, 48)],
+        names=["Z", "A"],
+    )
+
+    isotope_mass_fractions.columns = artis_model["cell_id"]
+    isotope_mass_fractions.columns.name = "cell_id"
+
+    if legacy_return:
+        return time_of_model, velocity, mean_density
+    else:
+        return time_of_model, velocity, mean_density, isotope_mass_fractions
+
+
+def read_artis_mass_fractions(fname, normalize=True):
+    """
+    Reads mass fractions from an artis abundance file named e.g. 'artis_abundances.dat'.
+    Each row typically corresponds to one shell. The first column is shell index,
+    followed by columns for each element.
+
+    Parameters
+    ----------
+    fname : str
+        filename or path to artis abundances
+
+    Returns
+    -------
+    mass_fractions : pandas.DataFrame
+        DataFrame containing mass fractions per shell (rows).
+        The first column is assumed to be shell index.
+    """
+    # Skip comment lines or set them to '#' if needed
+    # Here we assume no header row, so we'll assign column names ourselves
+    mass_fractions_df = pd.read_csv(
+        fname, comment="#", sep="\s+", header=None, index_col=0
+    )
+
+    mass_fractions_df.index.name = "cell_index"
+    mass_fractions_df.columns.Z = "element"
+
+    if normalize:
+        mass_fractions_df = mass_fractions_df.div(mass_fractions_df.sum(axis=1), axis=0)
+    mass_fractions_df = mass_fractions_df.T
+    mass_fractions_df.index.name = "Z"
+    mass_fractions_df.columns.name = "cell_id"
+
+    return mass_fractions_df
+
+
+def read_artis_model(density_fname, abundance_fname):
+    """
+    Reads both the density and abundance files and returns combined data.
+
+    Parameters
+    ----------
+    density_fname : str
+        Path to the ARTIS density file
+    abundance_fname : str
+        Path to the ARTIS abundance file
+
+    Returns
+    -------
+    ArtisModelData
+        Contains time_of_model, velocity, mean_density, and mass_fractions
+    """
+    time_of_model, velocity, mean_density, isotope_mass_fractions = read_artis_density(
+        density_fname, legacy_return=False
+    )
+    mass_fractions = read_artis_mass_fractions(abundance_fname)
+    mass_fractions.index = pd.MultiIndex.from_arrays(
+        [mass_fractions.index, [-1] * len(mass_fractions.index)], names=["Z", "A"]
+    )
+    isotope_summed = isotope_mass_fractions.groupby(level="Z").sum()
+    mass_fractions = mass_fractions.sub(isotope_summed, level="Z", fill_value=0)
+
+    mass_fractions = pd.concat([mass_fractions, isotope_mass_fractions], axis=0)
 
-    return time_of_model, velocity, mean_density
+    # Build the dataclass with the combined info:
+    return ArtisModelData(
+        time_of_model=time_of_model,
+        velocity=velocity,  # or keep as Quantity if you prefer
+        mean_density=mean_density,
+        mass_fractions=mass_fractions,
+    )
+from tardis.model.geometry.radial1d import HomologousRadial1DGeometry
```

Comment:
Import at end of file?

Reply:
to avoid circular import - do you want me to fix?

Reply:
Yes please